composer/datasets/streaming/dataset.py

```diff
@@ -392,6 +391,16 @@ def __iter__(self) -> Iterator[Any]:
             Iterator[Any]: Each sample.
         """
         Thread(target=self.download, daemon=True).start()
-
-        for idx in self._iter_ids():
-            yield self[idx]
+        #self.download()
+        while self._batch_count < self.index.total_samples:
+            try:
+                idx = self.shuffle_sample(self._batch_count) if self.shuffle else self._batch_count
+                yield self[idx]
+                self._batch_count += 1
```

Comment:
I don't think counting batches for the state dict inside the dataset will be accurate for serialization due to dataloader prefetching. When prefetching, the pytorch dataloader will attempt to iterate through the dataset greedily (up to the prefetch factor), so the number of samples the dataset yielded will be >= the number of samples actually used for training.